src/daemon/haven.rs

```diff
@@ -0,0 +1,63 @@
+use bytes::Bytes;
+use earendil_crypt::{Fingerprint, IdentityPublic, IdentitySecret};
+use earendil_packet::crypt::OnionPublic;
+use serde::{Deserialize, Serialize};
+use stdcode::StdcodeSerializeExt;
+
+#[derive(Clone, Debug, Deserialize, Serialize)]
+pub struct HavenLocator {
+    identity_pk: IdentityPublic,
+    onion_pk: OnionPublic,
+    rendezvous_fingerprint: Fingerprint,
+    signature: Bytes,
+}
+
+impl HavenLocator {
+    pub fn new(
+        identity_sk: IdentitySecret,
+        onion_pk: OnionPublic,
+        rendezvous_fingerprint: Fingerprint,
+    ) -> HavenLocator {
+        let identity_pk = identity_sk.public();
+        let signable = HavenLocator {
+            identity_pk,
+            onion_pk,
+            rendezvous_fingerprint,
+            signature: Bytes::new(),
+        };
+
+        let signature = identity_sk.sign(&signable.stdcode());
```

Comment:
you should sign a keyed hash of the stdcode here, so that signatures for different purposes are different